[Doc Enhancement] Introduce MarkdownElement #2224

frankf-cgn opened the pull request:
First working draft to implement #2224. Result can be seen on `<ListsPage>`.

In order to test, `cd /docs`, run `npm install` and `npm run start`. Visit `http://localhost:3000/#/components/lists` and scroll down to "Headline 1"

oliviertassinari commented:
I have no idea if this better or not. @shaurya947 What do you thin?

shaurya947 commented:
I think this is a slightly better system than what we have today. Obviously, it is easier to write a docs page as a separate markdown file vs. using `<p>`, `<h1>` and `<Codeblock>` elements inside the source code. Moreover, using this approach would be more robust when we think of a multi-lingual docs site: #1673

oliviertassinari commented on docs/src/app/components/markdown-element.jsx:
```diff
@@ -0,0 +1,44 @@
+const React = require('react');
```
What do you think about using ES6 features for this new component?
For instance import instead of require?

oliviertassinari commented on docs/src/app/components/pages/components/lists.jsx:
```diff
@@ -15,6 +16,8 @@ const EditorInsertChart = require('svg-icons/editor/insert-chart');
 const FileFolder = require('svg-icons/file/folder');
 const MoreVertIcon = require('svg-icons/navigation/more-vert');
 import { SelectableContainerEnhance } from 'material-ui/hoc/selectable-enhance';
+require('github-markdown-css/github-markdown.css');
```
I think that this should be move to `markdown-element.jsx`

frankf-cgn replied:
your right, there it belongs!

oliviertassinari commented on docs/src/app/components/pages/components/lists.jsx:
```diff
@@ -98,6 +102,7 @@ export default class ListsPage extends React.Component {
   }
 
   render() {
+    console.log("Markdown loaded: %O", ListsMarkdown);
```
To remove?

frankf-cgn replied:
...and gone.

oliviertassinari commented on docs/package.json:
```diff
@@ -15,8 +15,11 @@
+    "hightlight.js": "^8.9.1",
     "history": "^1.11.1",
     "intl": "^1.0.1",
+    "marked": "^0.3.5",
```
I think that the `raw-loader` dependency is missing.

frankf-cgn replied:
I'll check out if the raw-loader is needed in production, otherwise, I would like to let the raw-load  staying under devDependencies.

oliviertassinari commented on docs/src/app/components/markdown/lists-markdown.md:
````diff
@@ -0,0 +1,31 @@
+## Subheadline
+
+Detailes description here. And some code:
+```javascript
````
From experience, npm works best with just `js`. Should we follow this conventio?

frankf-cgn replied:
If it works best we can do so! Both (js and javascript) are possible with highlight.js

oliviertassinari commented on docs/src/app/components/pages/components/lists.jsx:
```diff
@@ -766,6 +771,9 @@ export default class ListsPage extends React.Component {
         </CodeExample>
+        <MarkdownElement text={`# Headline
+          No more info`}/>
+        <MarkdownElement text={ListsMarkdown} />
```
Do we need this?

frankf-cgn commented:
@oliviertassinari All the stuff in `docs/src/app/components/pages/components/lists.jsx` was just there to illustrate the usage, I'm going to remove it. 
I think I have to write some doc on how to use the `<MarkdownElement>`, but for now, I don't no where to put it.
Another open point is the css/style used within the Element. It's github-style and not very "material-design". Some adjustments seem to be appropriate. I'll work on this within the next days.

shaurya947 commented:
@frankf-cgn I can see people using this component in their applications. So perhaps it makes sense to document it as a new component on its own page? @oliviertassinari what say?

frankf-cgn commented:
@shaurya947 Strictly taken, this is not a ui-component in "material-design". It's more like `<CodeBlock>`, `<ComponentDoc>` or `<ComponentInfo>`, so I'm not sure if this is a good fit.

shaurya947 commented:
@oliviertassinari what do you think about making this a documented component with its own page like other components?

@frankf-cgn looking forward to the style changes!

oliviertassinari commented:
> just spoke to @hai-cea and seems like it's not adherent to the Material Design spec to treat this like a regular MUI component.

I agree, let's keep this component for our doc.

frankf-cgn commented:
@shaurya947 @oliviertassinari After pulling my hair over the lint-rules i finally managed to get this linted.

I have done the following stuff:
* reorganize the paths for the markdown files
* removed the package github-markdown 
* added a mui-github-markdown.css based on github-markdown.css with adjusted fonts settings and heading-styles
* reworked parts ot the `<ListPage>` as a showcase on the usage of `<MarkdownElement>`

Just out of curiosity I played around with various possibilities on how to format the documentation of props in markdown. May  ask for some feedback of you both on that?

lightj commented:
NAVER - http://www.naver.com/
--------------------------------------------

ngk0106@naver.com 님께 보내신 메일 <Re: [material-ui] [Doc Enhancement] Introduce MarkdownElement #2224 (#2226)> 이 다음과 같은 이유로 전송 실패했습니다.

--------------------------------------------

받는 사람이 회원님의 메일을 수신차단 하였습니다. 


--------------------------------------------

oliviertassinari commented on docs/src/app/components/markdown-element.jsx:
```diff
@@ -0,0 +1,48 @@
+    const {text} = this.props;
+    const html = marked(text || '');
+
+/* eslint-disable */
```
Humm, what'is the issue?

frankf-cgn replied:
To allow html-tags within markdown (e.g. for definition lists) `dangerouslySetInnerHTML` is needed which is prevented in `.eslint`L74 no-danger.

oliviertassinari commented:
> After pulling my hair over the lint-rules

Sorry, but this made me laught. I have enforced more rules lately. I aim to follow at least `eslint:recommended` and may be https://github.com/rackt/eslint-config-rackt.

> May ask for some feedback of you both on that 

I will have a look :+1:

oliviertassinari commented on docs/package.json:
```diff
@@ -15,8 +15,10 @@
   },
   "dependencies": {
     "codemirror": "^5.5.0",
+    "hightlight.js": "^8.9.1",
```
`.js`?

oliviertassinari commented on docs/package.json:
```diff
@@ -15,8 +15,10 @@
+    "hightlight.js": "^8.9.1",
     "history": "^1.11.1",
     "intl": "^1.0.1",
+    "marked": "^0.3.5",
```
I think that we are missing the `style-loader` dependency

frankf-cgn commented:
@oliviertassinari The ".js" is correct. but I have had a typo in "high**t**light.js", that is now fixed and the style-loader is added, too.

oliviertassinari commented on docs/package.json:
```diff
@@ -15,13 +15,16 @@
     "react-router": "^1.0.0-rc1",
-    "react-swipeable-views": "^0.3.0"
+    "react-swipeable-views": "^0.3.0",
+    "style-loader": "0.13.0"
```
css-loader is missing too.

oliviertassinari commented:
I have this following issue with webpack

```
ERROR in ./~/highlight.js/lib/languages/autoit.js
Module parse failed: /Users/oliviertassinari/material-ui/docs/node_modules/highlight.js/lib/languages/autoit.js Maximum call stack size exceeded
You may need an appropriate loader to handle this file type.
```

frankf-cgn commented:
@oliviertassinari I have no idea, why these npm-modules are installed globally in my environment - have to check this. Sorry for the noise it generates.
Despite I cannot reproduce the error you get with highlight.js/autoit.js I have found several hints how to mitigate this if it occures. I have modified the webpack-config accordingly. Can you check it out if it works for you now?

oliviertassinari commented:
@frankf-cgn With npm@3 and node@5, it's working :+1:

oliviertassinari commented:
@frankf-cgn I'm gonna start working on improving the doc. I'm keeping this PR in head to see how we can improve it!

oliviertassinari commented:
@frankf-cgn I'm gonna add your `MarkdownElement` this is a great component!
And use it for https://github.com/callemall/material-ui/pull/2382.

oliviertassinari commented:
@frankf-cgn I'm done, could you have a look at the other PR?
I think that we are on the right track :grimacing:

frankf-cgn commented:
@oliviertassinari Thats awesome! I'll look into the other PR.
